File: trunk/core/src/main/java/org/runetranscriber/core/tengwar/TengwaRune.java
package org.runetranscriber.core.tengwar;

import org.runetranscriber.core.DefaultRune;
import org.runetranscriber.core.Rune;
import org.runetranscriber.core.RuneList;

/**
 * Provides an enumeration of letters in the Tengwar alphabet.
 */
public final class TengwaRune implements Rune
{
    /** Primary letters. */
    /** Rune. */
    public static final TengwaRune TINCO = new TengwaRune("TINCO", "metal");

    /** Rune. */
    public static final TengwaRune PARMA = new TengwaRune("PARMA", "book");

    /** Rune. */
    public static final TengwaRune CALMA = new TengwaRune("CALMA", "lamp");

    /** Rune. */
    public static final TengwaRune QUESSE = new TengwaRune("QUESSE", "feather");

    /** 5-8 */
    public static final TengwaRune ANDO = new TengwaRune("ANDO", "gate");

    /** Rune. */
    public static final TengwaRune UMBAR = new TengwaRune("UMBAR", "fate");

    /** Rune. */
    public static final TengwaRune ANGA = new TengwaRune("ANGA", "iron");

    /** Rune. */
    public static final TengwaRune UNGWE = new TengwaRune("UNGWE", "spider's web");

    /** 9-12 */
    public static final TengwaRune THULE = new TengwaRune("THULE", "spirit or wind");

    /** Rune. */
    public static final TengwaRune FORMEN = new TengwaRune("FORMEN", "north");

    /** Rune. */
    public static final TengwaRune HARMA = new TengwaRune("HARMA", "treasure or rage");

    /** Rune. */
    public static final TengwaRune HWESTA = new TengwaRune("HWESTA", "breeze");

    /** 13-16 */
    public static final TengwaRune ANTO = new TengwaRune("ANTO", "mouth");

    /** Rune. */
    public static final TengwaRune AMPA = new TengwaRune("AMPA", "hook");

    /** Rune. */
    public static final TengwaRune ANCA = new TengwaRune("ANCA", "jaws");

    /** Rune. */
    public static final TengwaRune UNQUE = new TengwaRune("UNQUE", "a hollow");

    /** 17-20 */
    public static final TengwaRune NUMEN = new TengwaRune("NUMEN", "west");

    /** Rune. */
    public static final TengwaRune MALTA = new TengwaRune("MALTA", "gold");

    /** Rune. */
    public static final TengwaRune NGOLDO = new TengwaRune("NGOLDO", "one of the Noldor");

    /** Rune. */
    public static final TengwaRune NGWALME = new TengwaRune("NGWALME", "torment");

    /** 21-24 */
    public static final TengwaRune ORE = new TengwaRune("ORE", "heart or inner mind");

    /** Rune. */
    public static final TengwaRune VALA = new TengwaRune("VALA", "angelic power");

    /** Rune. */
    public static final TengwaRune ANNA = new TengwaRune("ANNA", "gift");

    /** Rune. */
    public static final TengwaRune VILYA = new TengwaRune("VILYA", "air or sky");

    /** Additional letters 25-28 */
    public static final TengwaRune ROMEN = new TengwaRune("ROMEN", "east");

    /** Rune. */
    public static final TengwaRune ARDA = new TengwaRune("ARDA", "region");

    /** Rune. */
    public static final TengwaRune LAMBE = new TengwaRune("LAMBE", "tongue");

    /** Rune. */
    public static final TengwaRune ALDA = new TengwaRune("ALDA", "tree");

    /** 29-32 */
    public static final TengwaRune SILME = new TengwaRune("SILME", "starlight");

    /** Rune. */
    public static final TengwaRune SILME_NUQUERNA = new TengwaRune("SILME_NUQUERNA", "silme reversed");

    /** Rune. */
    public static final TengwaRune ESSE = new TengwaRune("ESSE", "sunlight or name");

    /** Rune. */
    public static final TengwaRune ESSE_NUQUERNA = new TengwaRune("ESSE_NUQUERNA", "aze reversed");

    /** 33-36 */
    public static final TengwaRune HYARMEN = new TengwaRune("HYARMEN", "south");

    /** Rune. */
    public static final TengwaRune HWESTA_SINDARINWA = new TengwaRune("HWESTA_SINDARINWA", "sindarin hwesta");

    /** Rune. */
    public static final TengwaRune YANTA = new TengwaRune("YANTA", "bridge");

    /** Rune. */
    public static final TengwaRune URE = new TengwaRune("URE", "heat");

    /** Shorthand for 'the'. */
    public static final TengwaRune THE = new TengwaRune("THE", "the");

    /** Shorthand for 'of'. */
    public static final TengwaRune OF = new TengwaRune("OF", "of");

    /** Shorthand for 'of the'. */
    public static final TengwaRune OF_THE = new TengwaRune("OF_THE", "of the");

    /** Vowels. */
    /** Rune. */
    public static final TengwaRune SHORT_CARRIER = new TengwaRune("SHORT_CARRIER", "short vowel carrier");

    /** Rune. */
    public static final TengwaRune LONG_CARRIER = new TengwaRune("LONG_CARRIER", "long vowel carrier");

    /** A. */
    public static final TengwaRune THREE_DOTS = new TengwaRune("THREE_DOTS", "three dots");

    /** A. */
    public static final TengwaRune THREE_UNDER_DOTS = new TengwaRune("THREE_UNDER_DOTS", "three under dots");

    /** A. */
    public static final TengwaRune CIRCUMFLEX = new TengwaRune("CIRCUMFLEX", "circumflex");

    /** E. */
    public static final TengwaRune ACUTE = new TengwaRune("ACUTE", "acute");

    /** I. */
    public static final TengwaRune DOT = new TengwaRune("DOT", "dot");

    /** I. */
    public static final TengwaRune UNDER_DOT = new TengwaRune("UNDER_DOT", "under dot");

    /** O. */
    public static final TengwaRune LEFT_CURL = new TengwaRune("LEFT_CURL", "left curl");

    /** O. */
    public static final TengwaRune LEFT_UNDER_CURL = new TengwaRune("LEFT_UNDER_CURL", "left under curl");

    /** U. */
    public static final TengwaRune TOP_CURL = new TengwaRune("TOP_CURL", "top curl");

    /** U. */
    public static final TengwaRune RIGHT_CURL = new TengwaRune("RIGHT_CURL", "right curl");

    /** Y. */
    public static final TengwaRune CHEVRON = new TengwaRune("CHEVRON", "chevron");

    /** Punctuation. */
    /** Rune. */
    public static final TengwaRune SPACE = new TengwaRune("SPACE", "space");

    /** Rune. */
    public static final TengwaRune COMMA = new TengwaRune("COMMA", "comma");

    /** Rune. */
    public static final TengwaRune PERIOD = new TengwaRune("PERIOD", "period");

    /** Rune. */
    public static final TengwaRune NEWLINE = new TengwaRune("NEWLINE", "newline");

    /** Tehtar. (diacritic marks) */
    public static final RuneList<TengwaRune> TEHTAR = new RuneList<TengwaRune>();

    static
    {
        TEHTAR.add(THREE_DOTS);
        TEHTAR.add(THREE_UNDER_DOTS);
        TEHTAR.add(CIRCUMFLEX);
        TEHTAR.add(ACUTE);
        TEHTAR.add(DOT);
        TEHTAR.add(UNDER_DOT);
        TEHTAR.add(LEFT_CURL);
        TEHTAR.add(LEFT_UNDER_CURL);
        TEHTAR.add(TOP_CURL);
        TEHTAR.add(RIGHT_CURL);
        TEHTAR.add(CHEVRON);
    }

    /** Ordinal count. */
    private static int ordinalCount = 0;

    /** Values. */
    private static RuneList<TengwaRune> VALUES;

    /**
     * @param number Rune number.
     * 
     * @return rune.
     */
    public final static TengwaRune valueOfNumber(final int number)
    {
        TengwaRune answer = null;

        final RuneList<TengwaRune> values = values();

        if ((0 < number) && (number <= values.size()))
        {
            answer = values.get(number - 1);
        }

        return answer;
    }

    /**
     * @return a copy of the list of rune values.
     */
    public static final RuneList<TengwaRune> values()
    {
        return new RuneList<TengwaRune>(VALUES);
    }

    /** Delegate. */
    private final DefaultRune delegate;

    /**
     * Construct this object.
     * 
     * @param name Name.
     * @param nameMeaning Name meaning.
     */
    private TengwaRune(final String name, final String nameMeaning)
    {
        final String displayName = name.toLowerCase();
        delegate = new DefaultRune(ordinalCount, name, displayName, nameMeaning);

        if (VALUES == null)
        {
            VALUES = new RuneList<TengwaRune>();
        }

        VALUES.add(this);
        ordinalCount++;
    }

    @Override
    public String displayName()
    {
        return delegate.displayName();
    }

    @Override
    public boolean equals(final Object object)
    {
        boolean answer = false;

        if (object == this)
        {
            answer = true;
        }
        else if (object == null)
        {
            answer = false;
        }
        else if (getClass() != object.getClass())
        {
            answer = false;
        }
        else
        {
            final TengwaRune another = (TengwaRune)object;

            answer = delegate.equals(another.delegate);
        }

        return answer;
    }

    /**
     * @return the display name.
     */
    public String getDisplayName()
    {
        String answer = name().toLowerCase();

        answer = answer.replaceAll("[_]", " ");

        return answer;
    }

    @Override
    public int hashCode()
    {
        return delegate.hashCode();
    }

    @Override
    public String name()
    {
        return delegate.name();
    }

    @Override
    public String nameMeaning()
    {
        return delegate.nameMeaning();
    }

    @Override
    public int ordinal()
    {
        return delegate.ordinal();
    }

    @Override
    public String toDebugString()
    {
        return delegate.toDebugString();
    }

    @Override
    public String toString()
    {
        return delegate.toString();
    }
}

File: webapp/src/main/js/TengwaRune.js
/*
 * Provides an enumeration of letters in the Tengwar alphabet.
 */
define(function()
{
    "use strict";
    var TengwaRune =
    {
        TINCO: "tinco",
        PARMA: "parma",
        CALMA: "calma",
        QUESSE: "quesse",
        ANDO: "ando",
        UMBAR: "umbar",
        ANGA: "anga",
        UNGWE: "ungwe",
        THULE: "thule",
        FORMEN: "formen",
        HARMA: "harma",
        HWESTA: "hwesta",
        ANTO: "anto",
        AMPA: "ampa",
        ANCA: "anca",
        UNQUE: "unque",
        NUMEN: "numen",
        MALTA: "malta",
        NGOLDO: "ngoldo",
        NGWALME: "ngwalme",
        ORE: "ore",
        VALA: "vala",
        ANNA: "anna",
        VILYA: "vilya",
        ROMEN: "romen",
        ARDA: "arda",
        LAMBE: "lambe",
        ALDA: "alda",
        SILME: "silme",
        SILME_NUQUERNA: "silmeNuquerna",
        ESSE: "esse",
        ESSE_NUQUERNA: "esseNuquerna",
        HYARMEN: "hyarmen",
        HWESTA_SINDARINWA: "hwestaSindarinwa",
        YANTA: "yanta",
        URE: "ure",
        THE: "the",
        OF: "of",
        OF_THE: "ofThe",
        GH: "gh",
        SHORT_CARRIER: "shortCarrier",
        LONG_CARRIER: "longCarrier",
        THREE_DOTS: "threeDots",
        THREE_UNDER_DOTS: "threeUnderDots",
        CIRCUMFLEX: "circumflex",
        ACUTE: "acute",
        DOT: "dot",
        UNDER_DOT: "underDot",
        LEFT_CURL: "leftCurl",
        LEFT_UNDER_CURL: "leftUnderCurl",
        TOP_CURL: "topCurl",
        RIGHT_CURL: "rightCurl",
        CHEVRON: "chevron",
        BAR: "bar",
        UNDER_BAR: "underBar",
        HYPHEN: "hyphen",
        SPACE: "space",
        COMMA: "comma",
        PERIOD: "period",
        NEWLINE: "newline",

        properties:
        {
            "tinco":
            {
                name: "TINCO",
                displayName: "tinco",
                nameMeaning: "metal",
                fontLetter: "1",
                phoneme: "t",
            },
            "parma":
            {
                name: "PARMA",
                displayName: "parma",
                nameMeaning: "book",
                fontLetter: "q",
                phoneme: "p",
            },
            "calma":
            {
                name: "CALMA",
                displayName: "calma",
                nameMeaning: "lamp",
                fontLetter: "a",
                phoneme: "ch",
            },
            "quesse":
            {
                name: "QUESSE",
                displayName: "quesse",
                nameMeaning: "feather",
                fontLetter: "z",
                phoneme: "k",
            },
            "ando":
            {
                name: "ANDO",
                displayName: "ando",
                nameMeaning: "gate",
                fontLetter: "2",
                phoneme: "d",
            },
            "umbar":
            {
                name: "UMBAR",
                displayName: "umbar",
                nameMeaning: "fate",
                fontLetter: "w",
                phoneme: "b",
            },
            "anga":
            {
                name: "ANGA",
                displayName: "anga",
                nameMeaning: "iron",
                fontLetter: "s",
                phoneme: "j",
            },
            "ungwe":
            {
                name: "UNGWE",
                displayName: "ungwe",
                nameMeaning: "spider's web",
                fontLetter: "x",
                phoneme: "g",
            },
            "thule":
            {
                name: "THULE",
                displayName: "thule",
                nameMeaning: "spirit or wind",
                fontLetter: "3",
                phoneme: "th",
            },
            "formen":
            {
                name: "FORMEN",
                displayName: "formen",
                nameMeaning: "north",
                fontLetter: "e",
                phoneme: "f",
            },
            "harma":
            {
                name: "HARMA",
                displayName: "harma",
                nameMeaning: "treasure or rage",
                fontLetter: "d",
                phoneme: "sh",
            },
            "hwesta":
            {
                name: "HWESTA",
                displayName: "hwesta",
                nameMeaning: "breeze",
                fontLetter: "c",
            },
            "anto":
            {
                name: "ANTO",
                displayName: "anto",
                nameMeaning: "mouth",
                fontLetter: "4",
                phoneme: "th",
            },
            "ampa":
            {
                name: "AMPA",
                displayName: "ampa",
                nameMeaning: "hook",
                fontLetter: "r",
                phoneme: "v",
            },
            "anca":
            {
                name: "ANCA",
                displayName: "anca",
                nameMeaning: "jaws",
                fontLetter: "f",
            },
            "unque":
            {
                name: "UNQUE",
                displayName: "unque",
                nameMeaning: "a hollow",
                fontLetter: "v",
            },
            "numen":
            {
                name: "NUMEN",
                displayName: "numen",
                nameMeaning: "west",
                fontLetter: "5",
                phoneme: "n",
            },
            "malta":
            {
                name: "MALTA",
                displayName: "malta",
                nameMeaning: "gold",
                fontLetter: "t",
                phoneme: "m",
            },
            "ngoldo":
            {
                name: "NGOLDO",
                displayName: "ngoldo",
                nameMeaning: "one of the Noldor",
                fontLetter: "g",
            },
            "ngwalme":
            {
                name: "NGWALME",
                displayName: "ngwalme",
                nameMeaning: "torment",
                fontLetter: "b",
                phoneme: "ng",
            },
            "ore":
            {
                name: "ORE",
                displayName: "ore",
                nameMeaning: "heart or inner mind",
                fontLetter: "6",
                phoneme: "r",
            },
            "vala":
            {
                name: "VALA",
                displayName: "vala",
                nameMeaning: "angelic power",
                fontLetter: "y",
                phoneme: "w",
            },
            "anna":
            {
                name: "ANNA",
                displayName: "anna",
                nameMeaning: "gift",
                fontLetter: "h",
                phoneme: "y",
            },
            "vilya":
            {
                name: "VILYA",
                displayName: "vilya",
                nameMeaning: "air or sky",
                fontLetter: "n",
            },
            "romen":
            {
                name: "ROMEN",
                displayName: "romen",
                nameMeaning: "east",
                fontLetter: "7",
                phoneme: "r",
            },
            "arda":
            {
                name: "ARDA",
                displayName: "arda",
                nameMeaning: "region",
                fontLetter: "u",
            },
            "lambe":
            {
                name: "LAMBE",
                displayName: "lambe",
                nameMeaning: "tongue",
                fontLetter: "j",
                phoneme: "l",
            },
            "alda":
            {
                name: "ALDA",
                displayName: "alda",
                nameMeaning: "tree",
                fontLetter: "m",
            },
            "silme":
            {
                name: "SILME",
                displayName: "silme",
                nameMeaning: "starlight",
                fontLetter: "8",
                phoneme: "s",
            },
            "silmeNuquerna":
            {
                name: "SILME_NUQUERNA",
                displayName: "silme nuquerna",
                nameMeaning: "silme reversed",
                fontLetter: "i",
                phoneme: "c",
            },
            "esse":
            {
                name: "ESSE",
                displayName: "esse",
                nameMeaning: "sunlight or name",
                fontLetter: "k",
                phoneme: "z",
            },
            "esseNuquerna":
            {
                name: "ESSE_NUQUERNA",
                displayName: "esse nuquerna",
                nameMeaning: "aze reversed",
                fontLetter: ",",
                phoneme: "z",
            },
            "hyarmen":
            {
                name: "HYARMEN",
                displayName: "hyarmen",
                nameMeaning: "south",
                fontLetter: "9",
                phoneme: "h",
            },
            "hwestaSindarinwa":
            {
                name: "HWESTA_SINDARINWA",
                displayName: "hwesta sindarinwa",
                nameMeaning: "sindarin hwesta",
                fontLetter: "o",
                phoneme: "wh",
            },
            "yanta":
            {
                name: "YANTA",
                displayName: "yanta",
                nameMeaning: "bridge",
                fontLetter: "l",
                phoneme: "i",
            },
            "ure":
            {
                name: "URE",
                displayName: "ure",
                nameMeaning: "heat",
                fontLetter: ".",
                phoneme: "u",
            },
            "the":
            {
                name: "THE",
                displayName: "the",
                nameMeaning: "the",
                fontLetter: "@",
                phoneme: "the",
            },
            "of":
            {
                name: "OF",
                displayName: "of",
                nameMeaning: "of",
                fontLetter: "W",
                phoneme: "of",
            },
            "ofThe":
            {
                name: "OF_THE",
                displayName: "of the",
                nameMeaning: "of the",
                fontLetter: "W?",
                phoneme: "of the",
            },
            "gh":
            {
                name: "GH",
                displayName: "gh",
                nameMeaning: "gh",
                fontLetter: "X",
                phoneme: "gh",
            },
            "shortCarrier":
            {
                name: "SHORT_CARRIER",
                displayName: "short_carrier",
                nameMeaning: "short vowel carrier",
                fontLetter: "`",
            },
            "longCarrier":
            {
                name: "LONG_CARRIER",
                displayName: "long_carrier",
                nameMeaning: "long vowel carrier",
                fontLetter: "~",
            },
            "threeDots":
            {
                name: "THREE_DOTS",
                displayName: "three dots",
                nameMeaning: "three dots",
                fontLetter: "D",
                phoneme: "a",
            },
            "threeUnderDots":
            {
                name: "THREE_UNDER_DOTS",
                displayName: "three under dots",
                nameMeaning: "three under dots",
                fontLetter: "\u00D1",
                phoneme: "a",
            },
            "circumflex":
            {
                name: "CIRCUMFLEX",
                displayName: "circumflex",
                nameMeaning: "circumflex",
                fontLetter: "?",
            },
            "acute":
            {
                name: "ACUTE",
                displayName: "acute",
                nameMeaning: "acute",
                fontLetter: "R",
                phoneme: "i",
            },
            "dot":
            {
                name: "DOT",
                displayName: "dot",
                nameMeaning: "dot",
                fontLetter: "T",
                phoneme: "e",
            },
            "underDot":
            {
                name: "UNDER_DOT",
                displayName: "under dot",
                nameMeaning: "under dot",
                fontLetter: "\u00C9",
                phoneme: "e",
            },
            "leftCurl":
            {
                name: "LEFT_CURL",
                displayName: "left curl",
                nameMeaning: "left curl",
                fontLetter: "Y",
                phoneme: "o",
            },
            "leftUnderCurl":
            {
                name: "LEFT_UNDER_CURL",
                displayName: "left under curl",
                nameMeaning: "left under curl",
                fontLetter: "\u00E5",
                phoneme: "o",
            },
            "topCurl":
            {
                name: "TOP_CURL",
                displayName: "top curl",
                nameMeaning: "top curl",
                fontLetter: "J",
                phoneme: "u",
            },
            "rightCurl":
            {
                name: "RIGHT_CURL",
                displayName: "right curl",
                nameMeaning: "right curl",
                fontLetter: "\u00E1",
                phoneme: "u",
            },
            "chevron":
            {
                name: "CHEVRON",
                displayName: "chevron",
                nameMeaning: "chevron",
                fontLetter: "Ù",
            },
            "bar":
            {
                name: "BAR",
                displayName: "bar",
                nameMeaning: "bar",
                fontLetter: "{",
                phoneme: "a",
            },
            "underBar":
            {
                name: "UNDER_BAR",
                displayName: "underBar",
                nameMeaning: "underBar",
                fontLetter: "\u00ED",
            },
            "hyphen":
            {
                name: "HYPHEN",
                displayName: "hyphen",
                nameMeaning: "hyphen",
                fontLetter: "\u00B7",
                phoneme: "-",
            },
            "space":
            {
                name: "SPACE",
                displayName: "space",
                nameMeaning: "space",
                fontLetter: " ",
                phoneme: " ",
            },
            "comma":
            {
                name: "COMMA",
                displayName: "comma",
                nameMeaning: "comma",
                fontLetter: "\u00B7",
                phoneme: ",",
            },
            "period":
            {
                name: "PERIOD",
                displayName: "period",
                nameMeaning: "period",
                fontLetter: "-",
                phoneme: ".",
            },
            "newline":
            {
                name: "NEWLINE",
                displayName: "newline",
                nameMeaning: "newline",
                fontLetter: "\n",
                phoneme: "\n",
            },
        },

        // Tehtar (diacritic marks)
        tehtar: [ "threeDots", "threeUnderDots", "bar", // a
        "dot", "underDot", // e
        "acute", // i
        "leftCurl", "leftUnderCurl", // o
        "rightCurl", "topCurl", // u
        "underBar", "circumflex", "chevron" // other
        ],

        findByFontLetter: function(fontLetter)
        {
            var answer;
            var values = TengwaRune.values();
            var properties = TengwaRune.properties;

            for (var j = 0; j < values.length; j++)
            {
                var rune = values[j];

                if (properties[rune].fontLetter === fontLetter)
                {
                    answer = rune;
                    break;
                }
            }

            return answer;
        },

        findByPhoneme: function(phoneme)
        {
            var answer;
            var values = TengwaRune.values();
            var properties = TengwaRune.properties;

            for (var j = 0; j < values.length; j++)
            {
                var rune = values[j];

                if (properties[rune].phoneme === phoneme)
                {
                    answer = rune;
                    break;
                }
            }

            return answer;
        },

        isPunctuation: function(rune)
        {
            return [ TengwaRune.SPACE, TengwaRune.COMMA, TengwaRune.PERIOD, TengwaRune.NEWLINE ].includes(rune) ||
                    [ " ", ",", ".", "\n" ].includes(rune);
        },

        values: function()
        {
            return Object.getOwnPropertyNames(TengwaRune.properties);
        },
    };

    if (Object.freeze)
    {
        Object.freeze(TengwaRune);
    }

    return TengwaRune;
});
